src/main/scala/beam/agentsim/agents/ridehail/MatchmakingUtils.scala

```diff
@@ -0,0 +1,394 @@
+    cost
+  }
+
+  def computeGreedyCost(trip: RideHailTrip, vehicle: VehicleAndSchedule): Double = {
+    val passengers = trip.requests.size + vehicle.getNoPassengers
+    val delay = trip.sumOfDelays
+    val maximum_delay = trip.upperBoundDelays
+    val cost = passengers + (1 - delay / maximum_delay.toDouble)
+    -1 * cost
+  }
+
+  def checkAngle(origin: Coord, dest1: Coord, dest2: Coord)(implicit services: BeamServices): Boolean = {
+    val crs = DefaultGeographicCRS.WGS84
+    //val crs = MGC.getCRS(services.beamConfig.beam.spatial.localCRS)
+    val orgWgs = services.geo.utm2Wgs.transform(origin)
+    val dst1Wgs = services.geo.utm2Wgs.transform(dest1)
+    val dst2Wgs = services.geo.utm2Wgs.transform(dest2)
+    //val crs = CRS.decode(services.beamConfig.beam.spatial.localCRS)
+    val calc = new GeodeticCalculator(crs)
+    val gf = new GeometryFactory()
+    val point1 = gf.createPoint(new Coordinate(orgWgs.getX, orgWgs.getY))
+    calc.setStartingGeographicPoint(point1.getX, point1.getY)
+    val point2 = gf.createPoint(new Coordinate(dst1Wgs.getX, dst1Wgs.getY))
+    calc.setDestinationGeographicPoint(point2.getX, point2.getY)
+    val azimuth1 = calc.getAzimuth
+    val point3 = gf.createPoint(new Coordinate(dst2Wgs.getX, dst2Wgs.getY))
+    calc.setDestinationGeographicPoint(point3.getX, point3.getY)
+    val azimuth2 = calc.getAzimuth
+    val degrees = azimuth2 - azimuth1
+    Math.abs(degrees) < 45.0
+  }
+
+  def checkDistance(r: MobilityRequest, schedule: List[MobilityRequest], searchRadius: Double): Boolean = {
+    schedule.foreach { s =>
+      if (GeoUtils.distFormula(r.activity.getCoord, s.activity.getCoord) <= searchRadius)
+        return true
+    }
+    false
+  }
+
+  def getRequestsWithinGeofence(v: VehicleAndSchedule, demand: List[CustomerRequest]) = {
+    // get all customer requests located at a proximity to the vehicle
+    v.geofence match {
+      case Some(gf) =>
+        val gfCenter = new Coord(gf.geofenceX, gf.geofenceY)
+        demand.filter(
+          r =>
+            GeoUtils.distFormula(r.pickup.activity.getCoord, gfCenter) <= gf.geofenceRadius &&
+            GeoUtils.distFormula(r.dropoff.activity.getCoord, gfCenter) <= gf.geofenceRadius
+        )
+      case _ => demand
+    }
+  }
+
+  def getNearbyRequestsHeadingSameDirection(v: VehicleAndSchedule, demand: List[CustomerRequest])(
+    implicit services: BeamServices
+  ): List[CustomerRequest] = {
+    val requestWithCurrentVehiclePosition = v.getRequestWithCurrentVehiclePosition
+    val center = requestWithCurrentVehiclePosition.activity.getCoord
+
+    if (requestWithCurrentVehiclePosition.tag == EnRoute) {
+      // if vehicle is EnRoute, then filter list of customer based on the destination of the passengers
+      val i = v.schedule.indexWhere(_.tag == EnRoute)
+      val mainTasks = v.schedule.slice(0, i)
+      demand
+        .filter(
+          r =>
+            mainTasks.foldLeft(false) { (acc, t) =>
```

Comment:
Do you want everything reviewed, or would an exists be more appropriate here? It seems that way to me.